File: 173. Binary Search Tree Iterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// O(1)time, O(N)space, 9.09%, ok
class BSTIterator {
private:
    stack<int> ret;
public:
    BSTIterator(TreeNode *root) {
        priority_queue<int> max_heap;
        queue<TreeNode*> q;
        if(root) q.push(root);
        while(q.size()){
            TreeNode* temp = q.front();
            q.pop();
            max_heap.push(temp->val);
            if(temp->left) q.push(temp->left);
            if(temp->right) q.push(temp->right);
        }
        while(max_heap.size()){
            ret.push(max_heap.top());
            max_heap.pop();
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !ret.empty();
    }

    /** @return the next smallest number */
    int next() {
        int num = ret.top();
        ret.pop();
        return num;
    }
};

// O(1)time, O(h)space, 79.17%, ok
class BSTIterator {
private:
    stack<TreeNode*> st;
public:
    BSTIterator(TreeNode *root) {
        find_left(root);
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !st.empty();
    }

    /** @return the next smallest number */
    // brilliant idea that only keeps O(height) space
    int next() {
        TreeNode* top = st.top();
        st.pop();
        if (top->right != NULL)
            find_left(top->right);
        return top->val;
    }

    /** put all the left child() of root */
    void find_left(TreeNode* root){
        TreeNode* p = root;
        while(p != NULL){
            st.push(p);
            p = p->left;
        }
    }
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */

File: 173. Binary Search Tree Iterator.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator(object):
    
    # stack: avg O(1) time, O(h) space
    # stack records all the left-braches nodes, when next() called, insert all left-brached nodes of poped 'node' from stack if 'node' has a right branch  
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stack = []
        while root:
            self.stack.append(root)
            root = root.left

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.stack
        
    def next(self):
        """
        :rtype: int
        """
        node = self.stack.pop()
        curr = node.right
        while curr:
            self.stack.append(curr)
            curr = curr.left
        return node.val
        

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())